File: rust/tests/leetcode/merge_two_lists.rs
#[test]
fn test_merge_two_lists() {
    use practice::leetcode::merge_two_lists::{ListNode, Solution};
    let list1 = ListNode::from_vec(vec![1, 2, 4]);
    let list2 = ListNode::from_vec(vec![1, 3, 4]);
    let r = Solution::merge_two_lists(list1, list2);
    let rr = if r.is_some() {
        r.unwrap().to_vec()
    } else {
        vec![]
    };
    let expected = vec![1, 1, 2, 3, 4, 4];
    assert_eq!(rr, expected);
}

File: python/leetcode/merge_two_lists.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __eq__(self, value):
        return self.val == value.val and self.next == value.next


class Solution:
    def mergeTwoLists(
        self, list1: Optional[ListNode], list2: Optional[ListNode]
    ) -> Optional[ListNode]:
        match (list1, list2):
            case (None, None):
                return None
            case (None, list) | (list, None):
                return list
            case (list1, list2):
                if list1.val < list2.val:
                    return ListNode(list1.val, self.mergeTwoLists(list1.next, list2))
                else:
                    return ListNode(list2.val, self.mergeTwoLists(list1, list2.next))
